wurst/event/DamageEvent.wurst

```diff
@@ -0,0 +1,437 @@
+                default
+
+    Each damage instance can have an id and/or an element:
+
+        setNextEventDamageFromCode()
+        setNextEventDamageId('A000')
+        setNextEventDamageElement(FIRE_ELEMENT)
+        UnitDamageTarget(source, target, amount, [...])
+
+    Using globally defined priorities for your map can help you easily understand and use a damage pipeline:
+
+        public constant DAMAGE_EVENT_PRIO_START     = 0 // Abort damage at this priority
+        public constant DAMAGE_EVENT_PRIO_RELATIVE  = 1 // Apply relative changes (multiplication and division)
+        public constant DAMAGE_EVENT_PRIO_ABSOLUTE  = 2 // Apply absolute changes (addition and subtraction)
+        public constant DAMAGE_EVENT_PRIO_SHIELD    = 4 // Apply reduction from "shield" effects  
+        public constant DAMAGE_EVENT_PRIO_FINAL     = 5 // Final damage (won't be changed at this stage)
+*/
+package DamageEvent
+
+import AbilityObjEditing
+import ClosureEvents
+import ObjectIdGenerator
+import OnUnitEnterLeave
+
+/* CONFIGURATION */
+
+/** If true, native abilities that deal damage are passed with DamageType SPELL.
+    Also keep in mind if true:
+        > Spell damage is converted into negative damage values without this system, so do not use any other damage detection method.
+        > Spell reduction abilities like Runed bracers or Elunes Grace have to be coded.
+        > Life Drain doesn't heal the caster.
+        > Locust Swarm needs its Damage Return Factor changed from 0.75 to -0.75.
+        > Mana Shield abilities need their Damage Absorbed % reduced to 0.00 (and the damage absorbtion part has to be coded)
+        > Anti-Magic Shell (amount-based) has to be coded.
+    If false, native abilities that deal damage are passed with DamageType ATTACK */
+@configurable constant DETECT_NATIVE_ABILITIES = false
+
+/** A damage instance can have an element. 
+    DAMAGE_ELEMENT_ATTACK is the defaut element added to any damage instances of DamageType ATTACK */
+@configurable public constant DAMAGE_ELEMENT_ATTACK = new DamageElement("Physical", colorA(223, 59, 33, 255))
+
+
+/* CONSTANTS */
+
+constant DETECT_NATIVE_ABILITIES_ID  = compiletime(ABIL_ID_GEN.next())// Used if DETECT_NATIVE_ABILITIES is true
+
+
+/* DAMAGE TYPE */
+
+/** Each damage instance has a DamageType associated to it.
+    ATTACK and SPELL are detected by the system.
+    CODE **must** be declared by the user using setNextEventDamageFromCode() before dealing damage  */
+public enum DamageType
+    ATTACK  // From auto attacks or native abilities if DETECT_NATIVE_ABILITIES is false (PHYSICAL)
+    SPELL   // From native abilities (NATIVE_ABILITY)
+    CODE    // From user code
+    NULLED  // For backward compatibility -> has no utility
+    UNKNOWN         
+
+
+/* DAMAGE ELEMENT */
+
+/** Each damage instance can have a DamageElement associated to it. */
+public class DamageElement 
+    // Would be nice if we could @configurable a class 
```

Comment:
?!

Reply:
DamageElement is a way to create categories for your damage instances. The settings of those categories are tied to the map's gameplay so being able to configure the class in some ways would be nice. For now, people can just extends from it.

Reply:
Yea okay. Comment shouldn't be part of PR though.
Regarding `DamageElement` I think it's a good thing, the naming and usage just wasn't intuitive to me.
Element also to me implies that an attack can maybe have multiple elements.
Perhaps `CustomDamageType` or similar would be better. Element could be fine with updated doc.
Regarding doc, I think I would slim it down a bit and move examples and extra stuff into stdlib doc page: https://wurstlang.org/stdlib.html

Looking good 👍 
I think you can get rid of the class and just use a tuple for it though?